### DISCUSSION TITLE
Timeline for dropping support for legacy zlib compatibility

### DISCUSSION THREAD

**Discussion by mtl1979** [zlib-ng/zlib-ng#1382] [Release Feedback]
I think we should drop legacy zlib compatibility in zlib-ng 2.2.0, and remove `z_size_t`, replacing all `unsigned long` variants of functions permanently with variants using `size_t`... Keeping the compatibility mode for too long will just complicate maintaining the source code and cause issues especially for Windows builds.

Some big projects like GNU have already decided to stop using legacy zlib library for their projects and replace it with alternatives like zstd. Some distributions have replaced legacy zlib with zlib-ng 2.0.x , so we should still support the compatibility mode in 2.1.x series to give them and the developers of the libraries that they distribute enough time to switch to zlib-ng's own API. 

This means zlib-ng 2.1.x series must freeze the API early enough so that developers of third-party libraries don't have to resynchronise with API changes and can focus again on developing their own library after the initial switch.

**Comment by KungFuJesus** [comment]
Eh, I have some mixed feelings about this one.  There are a _lot_ of projects that use zlib.  It's an uphill battle to replace every call with a zlib-ng call.  One of the more compelling features of zlib-ng is that it's a direct, API compatible replacement for zlib.  You can usually (with some caveats) LD_PRELOAD the library into the loader and use it everywhere.  Your distro provides a slow libpng or chromium?  LD_PRELOAD it.  Some proprietary package that you have no access to the source to?  LD_PRELOAD it (or copy in the shared library to the working path with Windows).

Exactly what is the maintenance burden of having the compatibility shims in place?  The only notable one I have really noticed is long/size_t/long long type confusion (where a lot of platforms/operating systems define _what_ this type is differently, sometimes depending on the ISA).  I think the compatibility mode of zlib-ng is a pretty attractive feature and it'd lose a lot of charm if it lost this.

**Reply by mtl1979** [reply]
Some functions have two versions and some functions have potential overflow issues under platforms where `unsigned long` is 32-bit but `size_t` is 64-bit type. Currently we only support up to 2 GB files in compatibility mode even if the underlying file system and operating system support larger files.

In some cases we can have a shim that converts between pointers and non-pointer integer types before and after calling the internal function, but in some cases we would need two different versions even for compatibility mode due to being unable to change function prototypes in `zlib.h` without breaking API compatibility, or filing pull request or issue on https://github.com/madler/zlib repository and waiting for Mark Adler to respond, fix and make a new release.

**Reply by KungFuJesus** [reply]
Right, I get that poor ambiguous types were chosen for lengths for a lot of those functions.  And for any downstream consumer of zlib, those very same limitations would have applied (I would not be shocked if a lot of the autoconf/automake logic to check for sizeof(long) is strictly to sub in a different type at compile time to address these things).  

It's also highly likely that anyone running into these issues is well versed in them because they are trying to maintain legacy 32 bit binaries.  Arguably, 32 bit windows users are the largest percentage affected by this as Microsoft chose that silly convention of sizeof(long) == 4 on 32 bit variants.  There are others, and Microsoft is obviously not the sole offender of this, but I imagine that's almost everybody that would encounter this.  That user base is slowly but surely shrinking and those that still remain would have had this issue under the original API (this is not a compatibility layer specific bug).

I don't think throwing out all zlib compatibility is really a good logical conclusion to the snafu of choosing an ambiguous type for size in the API.  I think that would alienate a large percentage of the users that rely on the ability to swap between APIs at runtime or even just don't want to litter their code with macros at compile time.  Libraries like libjpeg-turbo still maintain API compatibility with libjpeg and treat libjpeg as a reference even decades after being adopted enough to be the de facto standard.  It'd be a good position to be in to be effectively upstream and not be dictated by zlib's API choices, but at the same time it's a big ask.

**Reply by mtl1979** [reply]
It's still too early to say when zlib-ng 2.2.0 will be released, but even zlib-ng 2.1 will be eventually supported for quite many years and it will still have the compatibility layer.

It's not really just about throwing away the compatibility layer, but also about end-of-life of legacy zlib itself... There might be zlib 2.x or something like that in the future, but that has nothing to do with compatibility layer for the current zlib 1.x series. zlib 1.2.12 already broke API compatibility with earlier zlib versions.

**Reply by KungFuJesus** [reply]
I think chasing binary compatibility with the current version of zlib at the time of release is a worthwhile goal. Trying to work with an older version of zlib is a good a time as any to shuck off support to an older variant of zlib-ng. 

I am one of probably many that would prefer not to have to patch every piece of code they want to make use of the latest version of zlib-ng.

**Reply by mtl1979** [reply]
What I've seen today, even new distributions are still using zlib 1.2.11... They haven't switched over to 1.2.12 or newer releases. Original zlib 1.2.12 was a dud release as it failed to build shared libraries on Linux. Mark Adler pretty much resumed work on zlib after hiatus lasting about 7 years... During that time, zlib-ng was started. 

Currently we try to be binary compatible with latest development version of zlib 1.2.12, so any project or distribution that used/use API of zlib 1.2.11 or older won't be able to use zlib-ng as replacement at all as the crc32 functions were rewritten from scratch.

**Reply by KungFuJesus** [reply]
Sure and for those, they can use an older release of zlib-ng to be compatible with that version. I think chasing the current interface of zlib at the time of development makes sense.

FWIW, every recent distro I've used and even illumos has switched to 1.2.12.

**Reply by mtl1979** [reply]
I just installed very latest version of Ubuntu (Lunal Lobster) and it says `1.2.11.dfsg-4.1ubuntu1`.

**Reply by KungFuJesus** [reply]
Fedora 37: 1.2.12: https://distrowatch.com/table.php?distribution=fedora&pkglist=true&version=37#pkglist
Arch Linux: 1.2.13: https://archlinux.org/packages/core/x86_64/zlib/
Gentoo: 1.2.13: https://packages.gentoo.org/packages/sys-libs/zlib
Debian testing (usual basis for Ubuntu, depending on release time): 1.2.13: https://distrowatch.com/table.php?distribution=debian&pkglist=true&version=testing#pkglist

It seems like Ubuntu is the weird holdout in this case.  I'm guessing because canonical has packages or is trying to maintain binary compatibility with something that required the older API.

**Reply by mtl1979** [reply]
Out of all the operating systems I have installed in my build cluster, a few use zlib 1.2.11 or older as default, few have some version of zlib-ng as default and the rest have either neither or both zlib and zlib-ng. I don't think any of them has zlib 1.2.13 though...

**Reply by KungFuJesus** [reply]
I don't think 1.2.13 had an API breaking change.  From the looks of it, they added a compatibility shim for the old CRC combine function:
https://github.com/madler/zlib/commit/888b3da8deebe1b637259b62256512fea104d690

Probably some outcry from distros that relied on old headers and code that used the old version or from tons of users that are used to zlib having a nearly static API for so long that weren't very careful with how they used it.

**Comment by nmoinvaz** [comment]
Zlib-compatibility is one of the big selling points of zlib-ng in my opinion. With the `ZLIB_COMPAT` option we have the best of both worlds.

**Reply by mtl1979** [reply]
But it will also drag along any limitations and bugs/regressions from zlib that we might have already fixed in our own API. Our main focus should be compatibility with modern processors and operating systems, not support processors operating systems that most people haven't used in over 10 or 20 years.

I have nothing against submitting patches for compatibility issues as long as the patch is clean enough to get accepted, but like I said earlier, some patches depend on if and how zlib will fix the limitation or bug in timely fashion.

**Comment by thesamesam** [comment]
This discussion has got me a bit worried about whether I should invest the time in making Gentoo compatible with zlib-ng (and I'm not the only distribution person who feels this way).

I don't think in Gentoo that we're ever going to be in a place where it's acceptable for us to drop legacy zlib binary compatibility (which implies the same behaviour & API compatibility, of course) entirely. If it's on the cards to drop the compatibility side entirely, I don't think it's worth me pursuing.

zlib-ng is attractive because it supports both the zlib interface and its own API, allowing people to upgrade to the improved API if desired, while not breaking compatibility with the 2 decades' worth of applications written for zlib. A lot of applications, if written today, wouldn't even use zlib at all. Losing compatibility with those would be a huge loss.

I think the value proposition of zlib-ng with 0 legacy compat. becomes pretty different, especially given most people understand zlib-ng as "zlib but with modern cleanups" rather than "zlib with no compatibility" (or "easy to port to").

Now, on the other hand, if you're just talking about bug-for-bug compatibility or similar with the latest zlib release at the time, then that's fine.

**Reply by Dead2** [reply]
I have no plans at all about dropping zlib compat mode.

Sure, it is a maintenance burden, and it would be nice if the world just converted to the zlib-ng native api, but that is unrealistic.
I can't promise that we won't drop it at some time in the future, but that would probably be after distros stop distributing the compat mode by default or something like that.

We might in a future release choose to move the compat code out of the main code, so it becomes a bit more of a wrapper to make maintenance easier. But that should not affect the API and users.

Bug-for-bug or byte-for-byte identical is something we do not advertise of course (It is much more interesting if we have our own bugs after all 😆).

**Reply by thesamesam** [reply]
Thanks for the reassurance! That's what I needed to hear, appreciate it. I'll bump up working on integration then :)

**Comment by hmaarrfk** [comment]
conda-forge core maintainer here, and long time lover of zlib (and now zlib-ng)

I'm curious if there is any new interest in this given the recent compatibility breakages with zlib (1.3.2) or if we should consider those exceptional in our plan to provide zlib-ng as a drop in replacement for zlib at conda-forge

your advise as the zlib-ng team on which direction we should go is appreciated

xref: https://github.com/conda-forge/conda-forge.github.io/issues/2638

**Reply by nmoinvaz** [reply]
I read the thread you referenced, there is a lot going on there.

zlib-ng has been adopted by several large open source projects and Linux distributions. It is used by Python 3.14 and .NET 10.

I don't think you should get hung up on supporting the latest zlib API - that is less than 6 months old and only involve a few new type changes to functions. A vast majority of users don't utilize them. And a few months is a short time in the lifespan of the entire project.

Zlib has not historically guaranteed full compatibility between its own versions. It is just that it hasn't been touched in such a long time until recently.

**Reply by nmoinvaz** [reply]
Also, @Dead2 has already mentioned that removing zlib compatibility support is unrealistic and he has no plans to do it. I also don't support removing it and I have said as much.

Regarding future API changes that zlib makes, those are always debatable - how, when, and if they are implemented. Anybody can make a PR if they notice a gap and that is the best place to handle such changes IMO.

**Reply by nmoinvaz** [reply]
In your decision you should also take into account the quality of the code in zlib repository which does not have automated code reviews, or an extensive CI stack, unit testing framework, or benchmarks run against it. For us, when incorporating changes from zlib, it becomes difficult because those changes do not always live up to the standards we subject our changes to.

**Reply by hmaarrfk** [reply]
> In your decision you should also take into account the quality of the code in zlib repository which does not have automated code reviews, or an extensive CI stack, unit testing framework, or benchmarks run against it. For us, when incorporating changes from zlib, it becomes difficult because those changes do not always live up to the standards we subject our changes to.


Thank you this is insightful.
